File: Heaps/minHeap/Heap.java
import java.util.NoSuchElementException;

/**
 * Abstract class for a Heap. 
 * A Heap is a binary tree following the rule, that a
 * new element is inserted at the first free space in
 * the tree and that each node is bigger than its parent
 * Node. If the second rule is violated after inserting
 * a new element, a "bubble-up" will be performed until
 * everything is okay.
 *
 * @author Thomas Lang
 * @version July 15, 2014
 * @see <a href="http://en.wikipedia.org/wiki/Binary_heap">Binary heap</a>
 */
public abstract class Heap<T extends Comparable<T>>{

    protected Node<T> root;
    protected int size;
    private static Object last; // pointer to last Node added

    /**
     * Constructor for a new Heap, initializes 
     * the <code>root</code> to <code>null</code>
     * and the <code>size</code> to 0.
     */
    public Heap(){
        root = null;
        size = 0;
        last = null;
    }

    /**
     * Checks, whether the heap is emtpy or not.
     *
     * @return <code>true</code> if the heap is empty, <code>false</code> otherwise
     */
    public boolean isEmpty(){
        return ( root == null );
    }

    /**
     * Gets the size of the Heap.
     *
     * @return The size of the heap.
     */
    public int getSize(){
        return size;
    }

    /**
     * Inserts a new Node to the heap.
     *
     * @param node  The node to insert
     */
    public void insert( Node<T> node ){
        if( isEmpty() ){
            root = node;
            size++;
            last = node;
        }else{
            insert( root, node );
            size++;
        }
    }

    /**
     * Inserts a new Node recursively into the Heap.
     * As this is just the abstract Heap, the node will
     * only be inserted, but not moved.
     *
     * @param  node     The current node.
     * @param  newNode  The new Node to insert.
     */
    private void insert( Node<T> node, Node<T> newNode ){
        if( node == null ){
            node = newNode;
            last = newNode;
        }else if( node.getLeft() == null ){
            node.setLeft( newNode );
            newNode.setParent( node );
            last = newNode;
        }else if( node.getRight() == null ){
            node.setRight( newNode );
            newNode.setParent( node );
            last = newNode;
        }else if( isFullTree( node ) && isFullTree( node.getLeft() )){
            insert( node.getRight(), newNode );
        }else if( isFullTree( node ) ){
            insert( node.getLeft(), newNode );
        }
    }

    /**
     * Helper function for checking if a node has both
     * a left and a right child.
     *
     * @param  node  The node to check.
     * @return <code>true</code> if the node has both children, <code>false</code> otherwise
     */
    private boolean isFullTree( Node<T> node ){
        if(( node != null ) && ( node.getLeft() != null ) && ( node.getRight() != null ))
                return true;
        return false;
    }

    /**
     * Gets the last Node inserted into the Heap.
     *
     * @return The last Node inserted into the Heap.
     * @throws NoSuchElementException
     */
    protected Node<T> getLastNode() throws NoSuchElementException{
        if( isEmpty() )
            throw new NoSuchElementException( "Error! No elements in Heap." );
        else
            return (Node<T>)last;
    }

    
    /**
     * Printing a Heap to stdout.
     */
    public void print(){
        if( isEmpty() )
            System.out.println( "(empty tree)" );
        System.out.println( "Root: " + root );
        printTreeRec( root );
    }

    /**
     * Recursive helper function for getting an appropiate
     * String representation of a Heap.
     *
     * @param  node  Current node.
     */
    private void printTreeRec( Node<T> node ){
        if( node.getLeft() != null )
            System.out.println( "Left: " + node.getLeft() + " "  );
        if( node.getRight() != null )
            System.out.println( "Right: " + node.getRight() );
        if( node.getLeft() != null )
            printTreeRec( node.getLeft() );
        if( node.getRight() != null )
            printTreeRec( node.getRight() );
    }
}

File: Heaps/maxHeap/Python/Heap.py

from Node import *

class Heap( object ):
    """Class representing a typical (abstract) Heap.
       This is principally a binary search tree that
       follows special rules. These differ on the 
       inheritated forms (MinHeap / MaxHeap), but a
       general rule is that new Nodes are inserted 
       strictly from left to right, there is NO 
       NUMERICAL connection between a left and a right
       Node."""

    def __init__( self ):
        """Creates a new but empty Heap."""
        self.root = None
        self.last = None  # pointer to last node added
        self.size = 0

    ################################################################################

    def isEmpty( self ):
        """Checks whether the Heap is empty or not.
        
        @return True if Heap is empty, False otherwise."""
        return ( self.root is None )

    ################################################################################

    def getSize( self ):
        """Returns the size (the number of 
           Nodes) of the Heap.
           
        @return The size of the Heap."""
        return self.size

    ################################################################################

    def insert( self, node ):
        """Inserts a new Node into the tree by
           following the rule mentioned above.
           
        @param  node  The node to insert.      """
        if self.isEmpty():
            self.root = node
            self.last = node
        else:
            self.__insert( self.root, node )
        self.size += 1


    def __insert( self, node, newNode ):
        """Helper method for inserting a
           Node recursively.
           
        @param  node     Current Node (for recursion)
        @param  newNode  The new Node to insert.     """
        if node is None:
            node = newNode
            self.last = newNode
        elif node.left is None:
            node.setLeft( newNode )
            newNode.setParent( node )
            self.last = newNode
        elif node.right is None:
            node.setRight( newNode )
            newNode.setParent( node )
            self.last = newNode
        elif self.__isFullTree( node ) and self.__isFullTree( node.left ):
            self.__insert( node.right, newNode )
        elif self.__isFullTree( node ):
            self.__insert( node.left, newNode )


    def __isFullTree( self, node ):
        """Helper function for checking whether the passed
           Node has both a left and a right child.
           
        @param  node  The node to check.
        @return True if the node has both a left and a right
                child, False otherwise"""
        if node is not None and node.left is not None and node.right is not None:
            return True
        return False

    ################################################################################
    
    def getLastNode( self ):
        """Returns a reference to the last Node added.
        
        @return A pointer to the last Node added.     """
        if self.isEmpty():
            raise LookupError( "Error! No elements in Heap." )
        else:
            return self.last

    ################################################################################

    def println( self ):
        """Prints the calling Heap to stdout."""
        if self.isEmpty():
            print( "(empty heap)" )
        else:
            print( "Root: " + str( self.root ) )
            self.__printRec( self.root )


    def __printRec( self, node ):
        """Helper function for printing.
        
        @param  node  Current node (for recursion). """
        if node.left is not None:
            print( "Left: " + str( node.left ))
        if node.right is not None:
            print( "Right: " + str( node.right ))
        if node.left is not None:
            self.__printRec( node.left )
        if node.right is not None:
            self.__printRec( node.right )

